File: src/uploadVideo.cpp
//
// Created by Xander Voorvaart on 15/05/2023.
//

#include <cstdlib> // for std::system
#include <sstream> // for std::stringstream
#include <filesystem>
#include <climits>
#include <unistd.h>
#include <string>

inline std::string getExecutablePath() {
	char result[PATH_MAX];
	ssize_t count = readlink("/proc/self/exe", result, PATH_MAX);
	std::string execPath = std::string( result, (count > 0) ? count : 0 );

	std::string::size_type pos = execPath.find_last_of("/\\");
	execPath = execPath.substr(0, pos);

	return execPath;
}


void callPythonScript(const std::filesystem::path& videoPath, const std::string& assetId) {
	std::string scriptPath = "../src/python/transcodeVideo.py";

	std::stringstream command;
	command << "python3 " << scriptPath << " " << videoPath << " " << assetId;

	int32_t result = std::system(command.str().c_str());

	if (result != 0) {
		throw std::runtime_error("Python script ended with error");
	}
}

File: src/upload_video.rs
use std::io;
use std::path::Path;
use std::process::Command;
use std::str;

pub fn call_python_script(video_path: &Path, asset_id: &str) -> io::Result<()> {
    let script_path = "/shorten/src/python/transcodeVideo.py";

    println!("Staring Python script");
    let output = Command::new("python3")
        .arg(script_path)
        .arg(video_path)
        .arg(asset_id)
        .output()?;

    if !output.status.success() {
        let stderr =
            str::from_utf8(&output.stderr).unwrap_or("Failed to parse python script stderr output");
        let stdout =
            str::from_utf8(&output.stdout).unwrap_or("Failed to parse python script stdout output");

        return Err(io::Error::new(
            io::ErrorKind::Other,
            format!(
                "Python script ended with error:\nStderr:\n{}\nStdout:\n{}",
                stderr, stdout
            ),
        ));
    }

    Ok(())
}
